### ISSUE TITLE
Throws error in node v8.0.0

### ISSUE THREAD

**Issue by jwhitmarsh** [archiverjs/node-archiver#236]
Code: 
```
const archive = archiver.create(`zip`, {});
archive.append(`test`, { name: `test.pdf` });
archive.finalize();

const writeStream = createWriteStream(`test.zip`);
await new Promise(function(resolve, reject) {
  writeStream.on(`close`, function() {
    console.log(archive.pointer() + ` total bytes`);
    console.log(`archiver has been finalized and the output file descriptor has closed.`);
    resolve();
  });
  writeStream.on(`error`, reject);

  archive.pipe(writeStream);
});
```

Throws:
```
_stream_readable.js:545
  switch (state.pipesCount) {
               ^

TypeError: Cannot read property 'pipesCount' of undefined
    at module.exports.Readable.pipe (_stream_readable.js:545:16)
    at module.exports.ZipArchiveOutputStream._smartStream (/Users/.../node_modules/compress-commons/lib/archivers/zip/zip-archive-output-stream.js:184:11)
    at module.exports.ZipArchiveOutputStream._appendBuffer (/Users/.../node_modules/compress-commons/lib/archivers/zip/zip-archive-output-stream.js:82:10)
    at module.exports.ArchiveOutputStream.entry (/Users/.../node_modules/compress-commons/lib/archivers/archive-output-stream.js:86:10)
    at module.exports.ZipStream.entry (/Users/.../node_modules/zip-stream/index.js:138:49)
    at Zip.append (/Users/.../node_modules/archiver/lib/plugins/zip.js:53:15)
    at Archiver._moduleAppend (/Users/.../node_modules/archiver/lib/core.js:172:16)
    at Archiver._onQueueTask (/Users/.../node_modules/archiver/lib/core.js:370:8)
    at /Users/.../node_modules/async/dist/async.js:4045:9
    at process (/Users/.../node_modules/async/dist/async.js:2316:17)
    at Immediate._onImmediate (/Users/.../node_modules/async/dist/async.js:66:16)
    at runCallback (timers.js:800:20)
    at tryOnImmediate (timers.js:762:5)
    at processImmediate [as _immediateCallback] (timers.js:733:5)
```

**Comment by ruslankonev** [comment]
me too

**Comment by soyuka** [comment]
More stacktrace? Weird thing is that it comes from [here](https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L545)

**Comment by jwhitmarsh** [comment]
@soyuka I've updated the issue with my stacktrace.

**Comment by soyuka** [comment]
Thanks @jwhitmarsh 

I can't explain the error though. [Archive output stream](https://github.com/archiverjs/node-compress-commons/blob/master/lib/archivers/archive-output-stream.js) correctly inherits the Transform stream which should have a `_readableState` unless node stream inheritance api has changed (AFAIR it has not - only thing added is the `destroy` method in node 8).

For the ease of debugging the source of the error is [zip-archive-output-stream (node-compress-commons)](https://github.com/archiverjs/node-compress-commons/blob/master/lib/archivers/zip/zip-archive-output-stream.js#L184). The stream there is using `node-crc32-stream` which may be part of the bug equation.

@ctalkington any idea on this?

**Comment by laurenzlong** [comment]
Same here.

**Comment by unipheas** [comment]
I'm not sure if this helps but here goes. I got this error trying to build an Ionic project with all the latest versions and got some extra information.

`(node:18146) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error [ERR_IPC_CHANNEL_CLOSED]: channel closed
(node:18146) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

**Comment by watilde** [comment]
One of the reasons why it doesn't work is this commit: https://github.com/nodejs/node/commit/8e69f7e38552a4c65064f2829d9ca973ad9b05ba. In v8.0.0, `zlib.DeflateRaw` is a class instead of a function, and we can't succeed it on the protype's way.
+ https://github.com/archiverjs/node-crc32-stream/blob/master/lib/deflate-crc32-stream.js#L13-L32

**Comment by soyuka** [comment]
Nice catch @watilde any idea on how this can be fixed? The best would be to have some kind of fallback for older nodejs versions :/.

Weird that this didn't show up in nodejs changelog btw.

**Comment by goelmk** [comment]
For now installed Node 6.10.3 LTS to bypass the issue for the time being until a fix for v 8.0.0 is available. It worked!

**Comment by jed** [comment]
i'm leaning on `zip` and `child_process` for now, ie:

before:
```js
var zip = archiver('zip').directory(pathname).finalize()
```
after:
```js
var zip = child_process.spawn('zip', ['-r', '-', pathname]).stdout
```

**Comment by ctalkington** [comment]
i'm following along with this and will work with node team as needed, seems crc32-stream may be added their testbed due to its popularity/dep usage to catch breakages earlier.

**Comment by dehimer** [comment]
https://github.com/nodejs/node/pull/13370

**Comment by watilde** [comment]
We may need to wait for this patch(https://github.com/nodejs/node/pull/13374) in Node.js core to fix this issue. Since the PR got some approved already, I think the patch will be landed in the next build(@jasnell is working on it https://github.com/nodejs/node/pull/13322#issuecomment-305676452).

As @ctalkington said, to not make this happen again, we're going to keep our eyes on the module with citgm: https://github.com/nodejs/citgm/pull/430.

**Comment by dikato** [comment]
same here using 'ionic upload', it broke this command.

**Comment by danielweck** [comment]
Same error with NodeJS v8 on TravisCI (I'm downgrading to NodeJS  v7 until this is fixed)

```
_stream_readable.js:545
  switch (state.pipesCount) {
               ^
TypeError: Cannot read property 'pipesCount' of undefined
    at module.exports.Readable.pipe (_stream_readable.js:545:16)
    at module.exports.ZipArchiveOutputStream._smartStream (/node_modules/compress-commons/lib/archivers/zip/zip-archive-output-stream.js:184:11)
    at module.exports.ZipArchiveOutputStream._appendStream (/node_modules/compress-commons/lib/archivers/zip/zip-archive-output-stream.js:96:20)
    at module.exports.ArchiveOutputStream.entry (/node_modules/compress-commons/lib/archivers/archive-output-stream.js:88:10)
    at module.exports.ZipStream.entry (/node_modules/zip-stream/index.js:138:49)
    at Zip.append (/node_modules/archiver/lib/plugins/zip.js:53:15)
    at Archiver._moduleAppend (/node_modules/archiver/lib/core.js:172:16)
    at Archiver._onQueueTask (/node_modules/archiver/lib/core.js:370:8)
    at /node_modules/async/dist/async.js:3879:9
```

**Comment by jasnell** [comment]
Yep. The patch already landed in core and 8.1 is coming out tomorrow with the fix.

**Comment by dehimer** [comment]
Cool, I thought its will be longer

**Comment by htoooth** [comment]
Same here.

**Comment by wayneashleyberry** [comment]
For anyone tracking this, here's a link to the 8.1.0 Proposal pull request.

https://github.com/nodejs/node/pull/13483

**Comment by unipheas** [comment]
Thanks!

**Comment by Bizarrus** [comment]
Same Issue here with node.js v8

**Comment by abenhamdine** [comment]
@Bizarrus Have you only checked nodejs/node#13483 and tried to install node 8.1.0 ?
And 'node.js v8' matches no existing version, it's either 8.0.0 or 8.1.0.

**Comment by soyuka** [comment]
Is resolved on 8.1.2
